unixfs/archive/archive.go

```diff
@@ -36,67 +36,69 @@ func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService
 
 	// need to connect a writer to a reader
 	piper, pipew := io.Pipe()
+	se := func(err error) bool {
+		if err != nil {
+			pipew.CloseWithError(err)
+			return true
+		}
+		return false
+	}
 
 	// use a buffered writer to parallelize task
 	bufw := bufio.NewWriterSize(pipew, DefaultBufSize)
 
 	// compression determines whether to use gzip compression.
-	var maybeGzw io.WriteCloser
-	var err error
-	if compression != gzip.NoCompression {
-		maybeGzw, err = gzip.NewWriterLevel(bufw, compression)
-		if err != nil {
-			pipew.CloseWithError(err)
-			return nil, err
+	maybeGzw, err := newMaybeGzWriter(bufw, compression)
+	if se(err) {
+		return nil, err
+	}
+
+	closePipe := func() {
+		maybeGzw.Close()
+		if err := bufw.Flush(); se(err) {
+			return
 		}
-	} else {
-		maybeGzw = &identityWriteCloser{bufw}
+		pipew.Close() // everything seems to be ok.
 	}
 
 	if !archive && compression != gzip.NoCompression {
 		// the case when the node is a file
 		dagr, err := uio.NewDagReader(ctx, nd, dag)
-		if err != nil {
-			pipew.CloseWithError(err)
+		if se(err) {
 			return nil, err
 		}
 
 		go func() {
-			if _, err := dagr.WriteTo(maybeGzw); err != nil {
-				pipew.CloseWithError(err)
-				return
-			}
-			maybeGzw.Close()
-			if err := bufw.Flush(); err != nil {
-				pipew.CloseWithError(err)
+			if _, err := dagr.WriteTo(maybeGzw); se(err) {
 				return
 			}
-			pipew.Close() // everything seems to be ok.
+			closePipe()
```

Comment:
`closePipe` is for the case "everything seems to be ok", `se` is for when err happens early. I defined the two funcs because they are common pattern for the 2 branch cases.